h/api/schemas.py

```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+"""Classes for validating data passed to the annotations API."""
+
+import jsonschema
+from jsonschema.exceptions import best_match
+
+
+# These annotation fields are not to be set by the user.
+PROTECTED_FIELDS = ['created', 'updated', 'user', 'id']
+
+
+class ValidationError(Exception):
+    pass
+
+
+class JSONSchema(object):
+
+    """
+    Validate data according to a Draft 4 JSON Schema.
+
+    Inherit from this class and override the `schema` class property with a
+    valid JSON schema.
+    """
+
+    schema = {}
+
+    def __init__(self):
+        self.validator = jsonschema.Draft4Validator(self.schema)
+
+    def validate(self, data):
+        """
+        Validate `data` according to the current schema.
+
+        :param data: The data to be validated
+        :return: valid data
+        :raises ValidationError: if the data is invalid
+        """
+        error = best_match(self.validator.iter_errors(data))
+        if error is not None:
+            raise ValidationError(_format_jsonschema_error(error))
+        return data
+
+
+class AnnotationSchema(JSONSchema):
+
+    """
+    Validate an annotation object.
+    """
+
+    schema = {
+        'type': 'object',
+        'properties': {
+            'document': {
+                'type': 'object',
+                'properties': {
+                    'link': {
+                        'type': 'array',
+                    },
+                },
+            },
+        },
+    }
+
+    def validate(self, data):
+        result = super(AnnotationSchema, self).validate(data)
+
+        # Some fields are not to be set by the user, ignore them
+        for field in PROTECTED_FIELDS:
+            result.pop(field, None)
+
+        return result
+
+
+def _format_jsonschema_error(error):
+    """Format a :py:class:`jsonschema.ValidationError` as a string."""
+    if error.path:
+        dotted_path = '.'.join(error.path)
+        return '{0}: {1}'.format(dotted_path, error.message)
```

Comment:
Note that `{1}` is not as good as `{error_message}` when a translator comes to localize a string. Of course this string will not be internationalized but it might be good for consistency reasons to always format strings in the same way everywhere?